packages/stash/src/actions/applyUpdates.ts

```diff
@@ -0,0 +1,91 @@
+import { schemaAbiTypeToDefaultValue } from "@latticexyz/schema-type/internal";
+import { Key, Stash, TableRecord, TableUpdates } from "../common";
+import { encodeKey } from "./encodeKey";
+import { Table } from "@latticexyz/config";
+import { registerTable } from "./registerTable";
+
+export type PendingStashUpdate<table extends Table = Table> = {
+  table: table;
+  key: Key<table>;
+  value: undefined | Partial<TableRecord<table>>;
+};
+
+export type ApplyUpdatesArgs = {
+  stash: Stash;
+  updates: PendingStashUpdate[];
+};
+
+type PendingUpdates = {
+  [namespaceLabel: string]: {
+    [tableLabel: string]: TableUpdates;
+  };
+};
+
+const pendingStashUpdates = new Map<Stash, PendingUpdates>();
+
+export function applyUpdates({ stash, updates }: ApplyUpdatesArgs): void {
+  const pendingUpdates = pendingStashUpdates.get(stash) ?? {};
+  if (!pendingStashUpdates.has(stash)) pendingStashUpdates.set(stash, pendingUpdates);
+
+  for (const { table, key, value } of updates) {
+    if (stash.get().config[table.namespaceLabel]?.[table.label] == null) {
+      registerTable({ stash, table });
+    }
+    const tableState = ((stash._.state.records[table.namespaceLabel] ??= {})[table.label] ??= {});
+    const encodedKey = encodeKey({ table, key });
+    const prevRecord = tableState[encodedKey];
+
+    // create new record, preserving field order
+    const nextRecord =
+      value == null
+        ? undefined
+        : Object.fromEntries(
+            Object.entries(table.schema).map(([fieldName, { type }]) => [
+              fieldName,
+              key[fieldName] ?? // Use provided key fields
+                value[fieldName] ?? // Or provided value fields
+                prevRecord?.[fieldName] ?? // Keep existing non-overridden fields
+                schemaAbiTypeToDefaultValue[type], // Default values for new fields
+            ]),
+          );
+
+    // apply update to state
+    if (nextRecord != null) {
+      tableState[encodedKey] = nextRecord;
+    } else {
+      delete tableState[encodedKey];
+    }
+
+    // add update to pending updates for notifying subscribers
+    const tableUpdates = ((pendingUpdates[table.namespaceLabel] ??= {})[table.label] ??= []);
+    tableUpdates.push({
+      table,
+      key,
+      previous: prevRecord,
+      current: nextRecord,
+    });
+  }
+
+  queueMicrotask(() => {
+    notifySubscribers(stash);
+  });
```

Comment:
```suggestion
  // TODO: wrap in queueMicrotask here once we figure out why it affects sync fallback
  notifySubscribers(stash);
```

Reply:
tested locally with a fresh template and everything seems fine

making a mental note to come back to this if we see weird issues with syncing, esp large chunks of data